packages/next-on-pages/templates/_worker.js/utils/fetch.ts

```diff
@@ -1,79 +1,121 @@
+ * 	https://github.com/vercel/next.js/blob/6705c803021d3bdea7fec20e5d98f6899e49836d/packages/next/src/server/web/sandbox/fetch-inline-assets.ts
+ *
+ * @param request the request to handle
+ * @returns the response to return to the caller if the request was for an inline asset one (and the file exists), null otherwise
+ */
+async function handleInlineAssetRequest(request: Request) {
+	if (request.url.startsWith('blob:')) {
+		try {
+			const url = new URL(request.url);
+			const binaryContent = (
+				await import(`./__next-on-pages-dist__/assets/${url.pathname}.bin`)
+			).default;
 
-				// Note: clone is necessary so that body does work
-				resp.clone = (): Response => {
-					return resp;
-				};
+			// Note: we can't generate a real Response object here because this fetch might be called
+			//       at the top level of a dynamically imported module, and such cases produce the following
+			//       error:
+			//           Some functionality, such as asynchronous I/O, timeouts, and generating random values,
+			//           can only be performed while handling a request
+			//       this is a somewhat known workerd behavior (currently kept for security and performance reasons)
+			//
+			//       if the above issue/constraint were to change we should replace the following with a real Response object
+			const resp = {
+				async arrayBuffer() {
+					return binaryContent;
+				},
+				get body(): ReadableStream<unknown> | null {
+					return new ReadableStream({
+						start(controller) {
+							return pump();
+							function pump() {
+								const b = Buffer.from(binaryContent);
+								controller.enqueue(b);
+								controller.close();
+								return pump();
+							}
+						},
+					});
+				},
+				async text() {
+					const b = Buffer.from(binaryContent);
+					return b.toString();
+				},
+				async json() {
+					const b = Buffer.from(binaryContent);
+					return JSON.stringify(b.toString());
+				},
+				async blob() {
+					return new Blob(binaryContent);
+				},
+			} as Response;
 
+			// Note: clone is necessary so that body does work
+			resp.clone = (): Response => {
 				return resp;
-			} catch {
-				/* empty, let's just fallback to the original fetch */
-			}
+			};
+
+			return resp;
+		} catch {
+			/* empty */
 		}
+	}
+	return null;
+}
 
-		return originalFetch(request);
-	};
+/**
+ *	updates the provided request by adding a Next.js specific user-agent header if the request has no user-agent header
+ *
+ *  Note: this is done by the Vercel network, but also in their next dev server as you can see here:
+ * 		https://github.com/vercel/next.js/blob/6705c803021d3bdea7fec20e5d98f6899e49836d/packages/next/src/server/web/sandbox/context.ts#L318-L320)
+ * @param request the request to update
+ */
+function setRequestUserAgentIfNeeded(
+	request: Request<unknown, RequestInitCfProperties>,
+): void {
+	if (!request.headers.has('user-agent')) {
+		request.headers.set(`user-agent`, `Next.js Middleware`);
+	}
```

Comment:
this is the only functional change in this PR, all the rest is just refactoring to make the code nicer